### ISSUE TITLE
Error calling StartServiceByName for org.freedesktop.portal.Desktop: Timeout was reached

### ISSUE THREAD

**Issue by mooreye** [Alexays/Waybar#2675]
This sometimes happens when launching Waybar for the first time since boot, subsequent Sway sessions do not then have this problem.

I have `xdg-desktop-portal`, `xdg-desktop-portal-gtk`, `xdg-desktop-portal-wlr` systemd user units, while `xdg-desktop-portal-gnome` is not installed.

I also have this in Sway config:

`exec dbus-update-activation-environment --systemd SWAYSOCK WAYLAND_DISPLAY XDG_CURRENT_DESKTOP XDG_SESSION_TYPE`

but it doesn't seem to help.

**Comment by RobertMueller2** [comment]
Have you tried this? https://github.com/emersion/xdg-desktop-portal-wlr/wiki/%22It-doesn't-work%22-Troubleshooting-Checklist

**Comment by mooreye** [comment]
WLR portal works, don't know what else to try really.
But I see discussion may be better place for this: https://github.com/Alexays/Waybar/discussions/1828
So maybe this issue can be closed.

**Comment by hsheikhali1** [comment]
I'm getting this same error on nixos and hyprland. When I try to start waybar it times out.

**Comment by mooreye** [comment]
It appears that masking WLR portal
`systemctl --user mask xdg-desktop-portal-wlr`
solved the issue, but I'm not 100% sure due to this occurring randomly beforehand.
Fortunately I don't need screen sharing, but all in all this is a poor solution

**Comment by SteelT1** [comment]
Same thing happens to me sometimes during startup
Since it was reported that masking the service for xdpw solved the issued, I'm going to try building xdpw from source and will report back.

As xdpw got a fix merged in it's polling code that would caused xdp to hang sometimes. https://github.com/emersion/xdg-desktop-portal-wlr/commit/b5f387821800e32d4f82001504668d819f02bb4b

Which would track of what I noticed that during the start delay, systemd reported xdp was still in the "activating" state, causing waybar to wait on it.

**Comment by svyatoclav** [comment]
Artix with runit + Hyprland so no systemd stuff ofc. `xdg-desktop-portal` and `xdg-desktop-portal-hyprland` installed. Same issue after `pacman -Syu` and reboot. I also changed some GTK + QT configs a little but I don't think it is a problem... I also move away my config directory for test but nothing. 

```
[2024-01-16 02:55:08.001] [info] Using configuration file //usr/local/etc/xdg/waybar/config
[2024-01-16 02:55:08.046] [error] Error calling StartServiceByName for org.freedesktop.portal.Desktop: Process org.freedesktop.portal.Desktop received signal 5
```

My waybar is built from source and was v0.9.22 and `meson_options.txt` for "Enable Experimental features" and "Disable SystemD" changes when it starts to happened (previously 3 months was OK but also sometimes it fails to launch at startup). Update to last `Waybar v0.9.24-102-g07eabc53 (branch 'master')` didn't fix. 
Anyone know what to do so far? Thanks.

**Comment by SteelT1** [comment]
> Same thing happens to me sometimes during startup Since it was reported that masking the service for xdpw solved the issue, I'm going to try building xdpw from source and will report back.
> 
> As xdpw got a fix merged in it's polling code that would caused xdp to hang sometimes. [emersion/xdg-desktop-portal-wlr@b5f3878](https://github.com/emersion/xdg-desktop-portal-wlr/commit/b5f387821800e32d4f82001504668d819f02bb4b)
> 
> Which would track of what I noticed that during the start delay, systemd reported xdp was still in the "activating" state, causing waybar to wait on it.

Okay it's been awhile, but I really wanted to make sure. And after a month of using this same setup, not once have I experience the start delay. So I'm pretty confident in that the fix to `xdg-desktop-portal-wlr` completely solved this issue for me.

**Comment by Archargelod** [comment]
Had the exact same issue with sway on OpenSuse Tumbleweed, solved it by removing the `xdg-desktop-portal-gtk` package.

**Comment by ouzbirki** [comment]
I'm having this problem on Arch Linux with Hyprland and none of the above worked.

<details>
<summary>Waybar logs</summary>

```
[2024-05-03 02:54:13.844] [debug] Try expanding: $XDG_CONFIG_HOME/waybar/config
[2024-05-03 02:54:13.844] [debug] Try expanding: $XDG_CONFIG_HOME/waybar/config.jsonc
[2024-05-03 02:54:13.844] [debug] Found config file: $XDG_CONFIG_HOME/waybar/config.jsonc
[2024-05-03 02:54:13.844] [info] Using configuration file /home/ouz/.config/waybar/config.jsonc
[2024-05-03 02:54:38.871] [error] Error calling StartServiceByName for org.freedesktop.portal.Desktop: Timeout was reached
```

</details>

<details>
<summary>Package details</summary>

```
Hyprland, built from branch main at commit a3ca016d42283ce1d2d5ffca40b93123b4290f4b  (cursor-shape: use - instead of _ in shapes).
Date: Thu May 2 18:21:33 2024
Tag: v0.39.1-128-ga3ca016d, commits: 4588
```

`Waybar v0.10.2-13-g0b6476da (branch 'master')`

`xdg-desktop-portal-gtk 1.15.2`

</details>

**Comment by squassina** [comment]
I was having this issue (Ubuntu 24.04 LTS aarch64) and what solved for me was to install xdg-desktop-portal-wlr and add this to my sway config (as per the xdg-desktop-portal-wlr README):

`exec dbus-update-activation-environment --systemd WAYLAND_DISPLAY XDG_CURRENT_DESKTOP=sway`

I was able to reboot and keep the bar. Hope this helps others.

**Comment by Its-Haze** [comment]
I get this error message from time to time on **Hyprland**. It sucks but i've gotten used to it.
Here is what i do:

Check if xdg-desktop-portal is running

```bash
systemctl --user status xdg-desktop-portal
```

If the service is not running, start it with:

```bash
systemctl --user start xdg-desktop-portal
```

Then start waybar again

---


Alternatively, you could run this one liner.

```bash
systemctl --user restart xdg-desktop-portal && systemctl --user daemon-reload && waybar &
```

**Comment by rickalex21** [comment]
I tried everything above, not sure what the issue is but on nixos waybar takes a while to start after trying some of the ideas on this post I did not have any luck. Any help would be appreciated.

```text
-❯ systemctl --user status xdg-desktop-portal
● xdg-desktop-portal.service - Portal service
     Loaded: loaded (/etc/systemd/user/xdg-desktop-portal.service; linked-runtime; preset: ignored)
     Active: active (running) since Wed 2024-12-11 17:45:41 CST; 11s ago
 Invocation: 5b4bf605ceea4c90a0e694444b573798
   Main PID: 12877 (.xdg-desktop-po)
      Tasks: 6 (limit: 18458)
     Memory: 2.5M (peak: 3.3M)
        CPU: 35ms
     CGroup: /user.slice/user-1000.slice/user@1000.service/session.slice/xdg-desktop-portal.service
             └─12877 /nix/store/w19i70qph6lfmbl55gbn55rs8j2d22hb-xdg-desktop-portal-1.18.4/libexec/xdg-desktop-portal

Dec 11 17:45:16 nixos systemd[1411]: Starting Portal service...
Dec 11 17:45:16 nixos .xdg-desktop-po[12877]: No skeleton to export
Dec 11 17:45:41 nixos .xdg-desktop-po[12877]: Failed to create screen cast proxy: Error calling StartServiceByName for org.freedesktop.impl.portal.desktop.wlr: Timeout was reached
Dec 11 17:45:41 nixos .xdg-desktop-po[12877]: No skeleton to export
Dec 11 17:45:41 nixos systemd[1411]: Started Portal service.
```

This is my nixos config:

```text
  systemd.user.services.xdg-desktop-portal-wlr = {
    environment.WAYLAND_DISPLAY = "wayland-1"; # Replace with your actual WAYLAND_DISPLAY value
 unitConfig = {
      ConditionEnvironment = "WAYLAND_DISPLAY";  # Adds the ConditionEnvironment under the Unit section
    };
  };
xdg.portal = {
  enable = true;
  extraPortals = with pkgs; [
    xdg-desktop-portal-wlr
    # xdg-desktop-portal-kde
    # xdg-desktop-portal-gtk
  ];
  wlr = {
    enable = true;
    settings = { # uninteresting for this problem, for completeness only
      screencast = {
        output_name = "DP-8";
        max_fps = 30;
        chooser_type = "simple";
        chooser_cmd = "${pkgs.slurp}/bin/slurp -f %o -or";
      };
    };
  };
    config.common.default = "*";
};

```

**Comment by fredrikv** [comment]
tl;dr: try adding the following to `~/.config/sway/config`:
```
exec systemctl --user import-environment DISPLAY WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP
exec hash dbus-update-activation-environment 2>/dev/null && \
        dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP=sway
```

I use Debian and had the same problem. The reason was that I had commented out the following line in `~/.config/sway/config`:
```
include /etc/sway/config.d/*
```

since the include caused `/etc/sway/config.d/40-sway-background.conf` to reset the background image. At the time it didn't seem like excluding the rest of the files in `config.d/` had any negative effects.

However, by moving that include up to the top of `~/.config/sway/config`, both `waybar` and my background images work properly.

What actually fixes the problem is including `/etc/sway/config.d/50-systemd-user.conf`. Here's the file content for anybody interested or using a different Linux distribution:
```
# import variables into system-user enviroment
# based on the instructions in the sway wiki
# see also https://github.com/swaywm/sway/issues/5732
# and https://github.com/systemd/systemd/blob/dfc637d0ff756889e8e5b7cb4ec991eb06069aa1/xorg/50-systemd-user.sh

exec systemctl --user import-environment DISPLAY WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP

exec hash dbus-update-activation-environment 2>/dev/null && \
        dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP=sway
```

These lines can of course be added directly to `~/.config/sway/config`.

Sharing this since the `dbus-update-activation-environment` command was not previously mentioned here, but I do not know under which circumstances it is needed.

**Comment by jmickelin** [comment]
For me, I had an issue not with `xdg-desktop-portal-wlr`, but with `xdg-desktop-portal-gnome`. It turns out that using the `GDK_DEBUG=portals` environment variable causes that portal, but not any others, to deadlock during login. I haven't investigated why, since I couldn't figure out how to monitor DBus calls during startup. 

Use of this flag is discouraged by the GTK people, but I really do want to use it (to prevent applications from using GTK UI components wherever possible). Turned out to be enough to tell systemd to unset it for just the Gnome portal service, which I did using a drop-in file.

Create `$HOME/.config/systemd/user/xdg-desktop-portal-gnome.service.d/overrides.conf`:
```service
[Service]
UnsetEnvironment=GDK_DEBUG
```

In case it matters, the way I define the variable for the rest of the user session is through `environment.d`:

```bash
$ cat $HOME/.config/environment.d/10-home-manager.conf
[...]
GDK_DEBUG=portals
[...]
```

**Comment by Alexays** [comment]
Closing — this was resolved/answered in the thread (config, environment, or packaging — not a Waybar bug still open). Feel free to reopen if you're still stuck. 🙏
